### PR TITLE
Stop truncating inclusion/exclusion circle radius to metre boundaries

### PR DESCRIPTION
**Pull request by peterbarker**
Just a casting issue

Tested round-tripping with QGC.

Also tested a little with autotest, but its infrastructure falls down when you need different epsilons for different fields.  It has a default epsilon of 1 - so any fractional changes are lost.  Will come back to that someday.

Replaces https://github.com/ArduPilot/MissionPlanner/issues/2631

Tests:
 - [x] Create fence with integer radius on this PR, ensure it reloads from eeprom
 - [x] Create fence with float radius on this PR, ensure it reloads from eeprom
 - [x] Create fence with integer radius on this PR, ensure it reloads from eeprom on Copter-4.0.7
 - [x] Create fence with integer radius on Copter-4.0.7, ensure it reloads from eeprom on this PR (floats truncated to integer)
 - [x] Create fence with float radius on this PR, ensure fences are completely empty on Copter-4.0.7 when loaded from eeprom

Note in that last test SITL actually dies with a SITL-specific "fence corrupt" message.  The autopilot also emits a message via statustext in this case.  When the panic is commented out the expected behaviour is present.

### CODE REVIEW

**Comment by rmackay9** [comment]
I've tested this and found that the radius parameter is changing as we move between versions.  So if a fence is written using "latest" and then read using this PR the Radius field becomes a very very large or very very small value.
image

If this can't be avoided then I'd be happy for the fences to be cleared when upgrading. I don't think there are many users of these complex fences yet even though I think they were introduced with Copter-4.0.

**Comment by peterbarker** [comment]
We could bump the version number to clear people's fences.  Fences would be cleared any time you move between old and new.

We could bump the version number and do a simple conversion to retain existing fences.

We could use a different item identifier to avoid bumping the version number - but you'd get weird things happening when going back to older versions (same as Randy saw, basically).

**Comment by tridge** [comment]
needs conversion code - possibly only convert on write?

**Comment by joshanne** [comment]
Do we need an explicit update to autotest to validate the changes as well? A test could be added to upload each of the point types and download the points and validate they are the same.

I'll accept we can't test across versions (upgrading to/downgrade from 4.0.x).

**Comment by peterbarker** [comment]
@joshanne it's difficult to tell without crossing versions what's going on inside, AFAICS.  The autotest already covers all of the externally-visible (mavlink) types.  How stuff is stored and magically converted should generally not be visible to autotest....

I was wondering about storing some old eeprom files in autotest that we could point a SITL instance at and check that the upgrade continues to work - but that's a large chunk of work.

Do you have any concrete suggestions on an autotest?  I wasn't able to think of one I was willing to implement while working on this :-)

**Comment by rmackay9** [comment]
I don't think we should overdo the autotests. Instead I think more general purpose tests that cover more situations is better than super detailed tests of a single bug we had in the past.

**Comment by IamPete1** [comment]
Just going to bump this, https://github.com/ArduPilot/ardupilot/pull/15550 it has some nice fence tests...

**Comment by joshanne** [comment]
I wasn't thinking new test, so much as perhaps extend something like fence static test (it's in plane right now). Could be extended to test circle inclusion and exclusion, and at the same time validate the circles that were uploaded are the same it downloaded.

**Review by rmackay9 — APPROVED** [review]
I've tested this PR on real hardware and it seems to work correctly.  My tests were:

1. Loaded Copter-4.0.7 onto the autopilot and created as list of fence points (circle exclusion, circle inclusion and polygon inclusion) with floating point radiuses and then confirmed that 4.0 could **not** correctly save/restore the radiuses (they were truncated as expected)
2. upgraded to this PR (4.1.0-dev+) and successfully downloaded the fence points to the GCS (with truncated radiuses)
3. changed radiuses to floating point values and confirmed they were saved and loaded correctly
4. downgraded to 4.0.7 and received a "Failed to get fence point" message displayed in MP and the fence area in the GCS was empty (which is fine)
image
5. upgraded to this PR again and the fence points were loaded correctly.

Along the way i found a Mission Planner bug (https://github.com/ArduPilot/MissionPlanner/issues/2635) but this shouldn't block this PR from being merged.

File: libraries/AC_Fence/AC_PolyFence_loader.h [hunk 1]
```diff
@@ -7,13 +7,21 @@
 
 #define AC_POLYFENCE_FENCE_POINT_PROTOCOL_SUPPORT 1
 
+// CIRCLE_INCLUSION_INT stores the radius an a 32-bit integer in
+// metres.  This was a bug, and CIRCLE_INCLUSION was created to store
+// as a 32-it float instead.  We save as _INT in the case that the
```
Comment (rmackay9) [inline libraries/AC_Fence/AC_PolyFence_loader.h:12]: small typo in the comments, "32-**b**it"

Comment (peterbarker) [reply libraries/AC_Fence/AC_PolyFence_loader.h:12]: Fixed.

**Pull request merged**